### PR TITLE
[SLP]Add GEP cost estimation for gathered loads.

### PR DESCRIPTION
**Pull request by alexey-bataev**
When doing estimation for vectorization of gathered loads, need to
estimate the cost of the pointer (vectorization), as it is done for the
actual vectorized loads. Otherwise may be too optimistic about the cost
of the gathered loads.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -7170,14 +7170,43 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
               TTI.getMemoryOpCost(Instruction::Load, LoadTy, Alignment,
                                   LI->getPointerAddressSpace(), CostKind,
                                   TTI::OperandValueInfo(), LI);
+          // Estimate GEP cost.
+          SmallVector<Value *> PointerOps(VF);
+          for (auto [I, V] : enumerate(VL.slice(P, VF)))
+            PointerOps[I] = cast<LoadInst>(V)->getPointerOperand();
+          InstructionCost ScalarGEPCost;
+          InstructionCost VectorGEPCost;
+          std::tie(ScalarGEPCost, VectorGEPCost) =
```
Comment (preames) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:7179]: You should be able to use:
auto [ScalarGEPCost, VectorGEPCost] = 

i.e. avoid forward declare and tie.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -7170,14 +7170,43 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
+          Align CommonAlignment = computeCommonAlignment<LoadInst>(Slice);
           GatherCost += TTI.getGatherScatterOpCost(
               Instruction::Load, LoadTy, LI0->getPointerOperand(),
               /*VariableMask=*/false, CommonAlignment, CostKind, LI0);
+          // Estimate GEP cost.
+          SmallVector<Value *> PointerOps(VF);
+          for (auto [I, V] : enumerate(Slice))
+            PointerOps[I] = cast<LoadInst>(V)->getPointerOperand();
+          OrdersType Order;
+          if (sortPtrAccesses(PointerOps, LI0->getType(), *R.DL, *R.SE,
```
Comment (preames) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:7196]: This bit of code is costing as-if we sort the accesses, but I don't see the cost of that shuffle being account for anywhere.  Am I missing something?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -7170,14 +7170,43 @@ class BoUpSLP::ShuffleCostEstimator : public BaseShuffleAnalysis {
...
+                              Order)) {
+            // TODO: improve checks if GEPs can be vectorized.
+            InstructionCost ScalarGEPCost;
+            InstructionCost VectorGEPCost;
+            Value *Ptr0 = PointerOps.front();
+            Type *ScalarTy = Ptr0->getType();
+            std::tie(ScalarGEPCost, VectorGEPCost) = getGEPCosts(
```
Comment (preames) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:7203]: Same style comment as above.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 2 above)
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:7196]: I'll add the cost for reordering here

**Review by preames — APPROVED** [review]
LGTM

I'm approving because the code change makes sense, the test delta on the example is actually a regression and you probably want to follow up on why.

**Pull request merged**